### ISSUE TITLE
Cannot try local docker version behind corporate firewall

### ISSUE THREAD

**Issue by herenyi** [talos-systems/talos#2261]
### Description
The command talosctl-linux-amd64 cluster create --wait fails as dns 8.8.8.8 is not available from corporate network.

### Logs
./talosctl-linux-amd64 cluster create --wait
validating CIDR and reserving IPs
generating PKI and tokens
downloading docker.io/autonomy/talos:v0.4.1
creating network talos-default
creating master nodes
creating worker nodes
waiting for etcd to be healthy: ...
waiting for etcd to be healthy: rpc error: code = Unavailable desc = connection error: desc = "transport: authentication handshake failed: EOF"
waiting for etcd to be healthy: rpc error: code = Unavailable desc = connection error: desc = "transport: authentication handshake failed: read tcp 127.0.0.1:36424->127.0.0.1:50000: read: connection reset by peer"
waiting for etcd to be healthy: rpc error: code = Unavailable desc = connection error: desc = "transport: authentication handshake failed: EOF"
waiting for etcd to be healthy: service "etcd" not in expected state "Running": current state [Waiting] Waiting for service "networkd" to be "up"
waiting for etcd to be healthy: service "etcd" not in expected state "Running": current state [Preparing] Running pre state
waiting for etcd to be healthy: service "etcd" not in expected state "Running": current state [Failed] Failed to run pre stage: failed to pull image "k8s.gcr.io/etcd:3.3.15-0": 3 error(s) occurred:
        failed to pull image "k8s.gcr.io/etcd:3.3.15-0": failed to resolve reference "k8s.gcr.io/etcd:3.3.15-0": failed to do request: Head "https://k8s.gcr.io/v2/etcd/manifests/3.3.15-0": dial tcp: lookup k8s.gcr.io on 8.8.8.8:53: read udp 10.5.0.2:38132->8.8.8.8:53: i/o timeout
        failed to pull image "k8s.gcr.io/etcd:3.3.15-0": failed to resolve reference "k8s.gcr.io/etcd:3.3.15-0": failed to do request: Head "https://k8s.gcr.io/v2/etcd/manifests/3.3.15-0": dial tcp: lookup k8s.gcr.io on 8.8.8.8:53: read udp 10.5.0.2:53751->8.8.8.8:53: i/o timeout
        timeout
context deadline exceeded

### Environment

- Talos version: [`talosctl version --nodes <problematic nodes>`]
- Kubernetes version: [`kubectl version --short`]
- Platform:

**Comment by rsmitty** [comment]
Hi @herenyi, there is a `--nameservers` flag in talosctl cluster create that might solve your issue. Ping me here or on our slack channel if it doesn't work for you.

**Comment by herenyi** [comment]
Hi, thanks for the tip, unfortunately it didn't work for me:
./talosctl-linux-amd64 cluster create --wait --nameservers 10.130.208.12,10.158.51.12
validating CIDR and reserving IPs
generating PKI and tokens
creating network talos-default
creating master nodes
creating worker nodes
waiting for etcd to be healthy: ...
waiting for etcd to be healthy: rpc error: code = Unavailable desc = connection error: desc = "transport: authentication handshake failed: read tcp 127.0.0.1:33722->127.0.0.1:50000: read: connection reset by peer"
waiting for etcd to be healthy: rpc error: code = Unavailable desc = connection error: desc = "transport: authentication handshake failed: read tcp 127.0.0.1:33742->127.0.0.1:50000: read: connection reset by peer"
waiting for etcd to be healthy: service "etcd" not in expected state "Running": current state [Waiting] Waiting for service "networkd" to be "up"
waiting for etcd to be healthy: service "etcd" not in expected state "Running": current state [Preparing] Running pre state
waiting for etcd to be healthy: service "etcd" not in expected state "Running": current state [Failed] Failed to run pre stage: failed to pull image "k8s.gcr.io/etcd:3.3.15-0": 3 error(s) occurred:
        failed to pull image "k8s.gcr.io/etcd:3.3.15-0": failed to resolve reference "k8s.gcr.io/etcd:3.3.15-0": failed to do request: Head "https://k8s.gcr.io/v2/etcd/manifests/3.3.15-0": dial tcp: lookup k8s.gcr.io on 8.8.8.8:53: read udp 10.5.0.2:55197->8.8.8.8:53: i/o timeout
        failed to pull image "k8s.gcr.io/etcd:3.3.15-0": failed to resolve reference "k8s.gcr.io/etcd:3.3.15-0": failed to do request: Head "https://k8s.gcr.io/v2/etcd/manifests/3.3.15-0": dial tcp: lookup k8s.gcr.io on 8.8.8.8:53: read udp 10.5.0.2:55883->8.8.8.8:53: i/o timeout
        timeout
context deadline exceeded

**Comment by rsmitty** [comment]
@herenyi thanks for verifying. I'll spend some time troubleshooting this today.

**Comment by rsmitty** [comment]
@herenyi mind telling me what version of talosctl you're using? it looks like that's deploying a pretty old version of talos, so you may want to pull the latest talosctl down from our releases and try it again with the nameservers flag. seems to be working for me, as I tried the router's DNS and cloudflare's DNS:

```
$ talosctl cluster create --nameservers 192.168.1.254
...
...
$ talosctl read /etc/resolv.conf
nameserver 192.168.1.254
```

```
$ talosctl cluster create --nameservers 1.1.1.1
...
...
$ talosctl read /etc/resolv.conf
nameserver 1.1.1.1
```

**Comment by herenyi** [comment]
I was using 0.4.1, now I have tried it with 0.5.0 and I think the nameserver works, but somehow it is not using the proxy or some other problems as it is showing now the following error:
./talosctl-linux-amd64 cluster create --wait --nameservers 10.130.208.12,10.158.51.12
validating CIDR and reserving IPs
generating PKI and tokens
creating network talos-default
creating master nodes
creating worker nodes
waiting for etcd to be healthy: ...
waiting for etcd to be healthy: rpc error: code = Unavailable desc = connection error: desc = "transport: authentication handshake failed: read tcp 127.0.0.1:55606->127.0.0.1:50000: read: connection reset by peer"
waiting for etcd to be healthy: rpc error: code = Unavailable desc = connection error: desc = "transport: authentication handshake failed: read tcp 127.0.0.1:55646->127.0.0.1:50000: read: connection reset by peer"
waiting for etcd to be healthy: rpc error: code = Unavailable desc = connection error: desc = "transport: authentication handshake failed: EOF"
waiting for etcd to be healthy: service "etcd" not in expected state "Running": current state [Preparing] Running pre state
waiting for etcd to be healthy: service "etcd" not in expected state "Running": current state [Failed] Failed to run pre stage: failed to pull image "k8s.gcr.io/etcd:3.3.15-0": 2 error(s) occurred:
        failed to pull image "k8s.gcr.io/etcd:3.3.15-0": failed to resolve reference "k8s.gcr.io/etcd:3.3.15-0": failed to do request: Head "https://k8s.gcr.io/v2/etcd/manifests/3.3.15-0": dial tcp 74.125.204.82:443: i/o timeout
        timeout

I have tried to pull manually the etcd image (which was successful), but it didn't help... any ideas how to proceed?
Thank you in advance!
Cheers,
Gergo

**Comment by smira** [comment]
@herenyi one of the options is to use [Registry Cache](https://www.talos.dev/docs/v0.6/en/guides/local/registry-cache) (docs are for 0.6, but it's available in 0.5)

if you can make registry containers to pull the images on your host, Talos will be able to download them from the registry cache.

**Comment by rsmitty** [comment]
Another possibility, are you setting your proxy options in the docker daemon? Looking at https://docs.docker.com/network/proxy/, it looks fairly straightforward and I wonder if that would fix your issue.

**Comment by rsmitty** [comment]
Another bit to try if that doesn't work would be to configure the proxies on talos itself. You can do that by pre-generating the talos configs, editing them with the proxy information, then doing a `talosctl cluster create`.

An example:

- Gen configs with ` talosctl gen config -o /desired/output/directory talos-default https://10.5.2.2:6443`
- Edit the each of the .yaml configs to add the proxy env vars as mentioned [here](https://www.talos.dev/docs/v0.5/en/customization/proxy)
- Create the cluster with `talosctl cluster create -i /desired/output/directory`

**Comment by smira** [comment]
yep, @rsmitty is exactly right, if HTTP proxy is configured, it should update the image pull machinery inside Talos (vs. docker itself)

**Comment by herenyi** [comment]
> 
> 
> @herenyi one of the options is to use [Registry Cache](https://www.talos.dev/docs/v0.6/en/guides/local/registry-cache) (docs are for 0.6, but it's available in 0.5)
> 
> if you can make registry containers to pull the images on your host, Talos will be able to download them from the registry cache.

Registry cache idea has worked, thanks.
